drivers/iommu/riscv/iommu.c

```diff
@@ -661,27 +771,60 @@ static void riscv_iommu_page_request(struct riscv_iommu_device *iommu,
 	if (req->payload & RISCV_IOMMU_PREQ_PAYLOAD_R)
 		prm->perm |= IOMMU_FAULT_PERM_READ;
 
-	prm->grpid = FIELD_GET(RISCV_IOMMU_PREQ_PRG_INDEX, req->payload);
+	resp.grpid = prm->grpid = FIELD_GET(RISCV_IOMMU_PREQ_PRG_INDEX, req->payload);
 	prm->addr = FIELD_GET(RISCV_IOMMU_PREQ_UADDR, req->payload);
 
 	if (req->hdr & RISCV_IOMMU_PREQ_HDR_PV) {
 		prm->flags |= IOMMU_FAULT_PAGE_REQUEST_PASID_VALID;
 		/* TODO: where to find this bit */
 		prm->flags |= IOMMU_FAULT_PAGE_RESPONSE_NEEDS_PASID;
-		prm->pasid = FIELD_GET(RISCV_IOMMU_PREQ_HDR_PID, req->hdr);
+		resp.flags |= IOMMU_PAGE_RESP_PASID_VALID;
+		resp.pasid = prm->pasid = FIELD_GET(RISCV_IOMMU_PREQ_HDR_PID, req->hdr);
 	}
 
-	iommu_report_device_fault(ep->dev, &event);
+	ret = iommu_report_device_fault(ep->dev, &event);
+	resp.code = (ret) ? IOMMU_PAGE_RESP_FAILURE : IOMMU_PAGE_RESP_SUCCESS;
+	// This is going to invoke riscv_iommu_page_response() under the hood
+	// because it's registered as the page_response() callback from the
+	// iommu_ops structure.
+	iommu_page_response(ep->dev, &resp);
+
```

Comment:
Ah I had totally missed that... But this is logical indeed, given we want to make sure to issue the page response after the actual page fault occurred.
I'm going to send an ATS.PRGR failure in case `iommu_report_device_fault()` returned an error.